daemon/daemon.go

```diff
@@ -805,6 +805,11 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
 	if err := d.setGenericResources(config); err != nil {
 		return nil, err
 	}
+
+	if err := initializeCdi(d); err != nil {
```

Comment:
`NewDaemon` is not the ideal place to put code which mutates global state and not the daemon struct. Package `./cmd/dockerd` would be a better fit for such global concerns. I think [`(*DaemonCli).start`](https://github.com/moby/moby/blob/2cf63891427d68661bb8e3c58969058c62dbc85d/cmd/dockerd/daemon.go#L78) would be the best place to register the CDI device driver since that function is responsible for loading the daemon config, preparing global state and instantiating the daemon object.

Reply:
OK. Will update with that in mind.

Reply:
I've just been looking at this and one question I have: Registering the device driver currently sets a package-global map in the `daemon` package. This means that the `InitializeCDI` function needs to be public. Is this ok, and if so, would we prefer a function on the `*Daemon` type, or a package-level function. If the latter is used, should we perform the experimental check in the `(*DaemonCli).start` function instead?

Another option would be to refactor the driver registration in preparation for this and move all the CDI code to the CLI.

Reply:
Exporting the `InitializeCDI` function is fine with me. It should not be associated with `*Daemon`, for the same reason it does not belong in the `NewDaemon` constructor: it affects global state, not a particular `Daemon` instance. Perform the experimental check in `(*DaemonCli).start`.

> Another option would be to refactor the driver registration in preparation for this and move all the CDI code to the CLI.

I would be happy to see (and review) a refactor of the driver registration code, but neither that nor the bulk of the CDI code belongs in package `cmd/dockerd`. Now, if we're talking refactoring, I would love to see the device-driver registration code and `handleDevice` function extracted to an object which is instantiated and dependency-injected into the daemon instance, like the `pluginStore`.

Reply:
While we're talking 4-year-old refactoring TODOs... https://github.com/moby/moby/blob/7489b51f610104ab5acc43f4e77142927e7b522e/daemon/nvidia_linux.go#L15

Reply:
From the NVIDIA perspective, we see the use of CDI replacing the `--gpus` flag in the long run. The pointers on doing this better this time around are much appreciated.

I will have a look at the `pluginStore` and take a pass at the refactoring in a follow-up!